fix(ngcc): detect non-emitted, non-imported TypeScript helpers

gkalpak opened the pull request:
When TypeScript downlevels ES2015+ code to ES5, it uses some helpers functions to emulated some ES2015+ features, such as spread syntax. The TypeScript compiler can be configured to emit these helpers into the transpiled code (which is controlled by the `noEmitHelpers` option - false by default). It can also be configured to import these helpers from the `tslib` module (which is controlled by the `importHelpers` option - false by default).

While most of the time the helpers will be either emitted or imported, it is possible that one configures their app to neither emit nor import them. In that case, the helpers could, for example, be made available on the global object. This is what `@nativescript/angular` v9.0.0-next-2019-11-12-155500-01 does. See, for example, [common.js][1].

Ngcc must be able to detect and statically evaluate these helpers. Previously, it was only able to detect emitted or imported helpers.

This commit adds support for detecting these helpers if they are neither emitted nor imported . It does this by checking identifiers for which no declaration (either concrete or inline) can be found against a list of known TypeScript helper function names.

##
This is needed for angular/ngcc-validation#1029.

[1]: https://unpkg.com/browse/@nativescript/angular@9.0.0-next-2019-11-12-155500-01/common.js

gkalpak commented:
Marking this as blocked on #36284, since the latter has to go in first.

mary-poppins commented:
You can preview 406536b at https://pr36418-406536b.ngbuilds.io/.

mary-poppins commented:
You can preview 3ecd4ca at https://pr36418-3ecd4ca.ngbuilds.io/.

petebacondarwin commented:
A couple of typos in the commit message:

`When TypeScript downlevels ES2015+ code to ES5, it uses some helper~~s~~`

`functions to emulate~~d~~ some ES2015+ features, such as spread syntax.`

`emitted nor imported~~ ~~. It does this by checking identifiers for which no`

petebacondarwin reviewed (approved):
LGTM save the typos in the commit message

JoostK reviewed (approved):
💯

gkalpak committed:
fix(ngcc): detect non-emitted, non-imported TypeScript helpers

When TypeScript downlevels ES2015+ code to ES5, it uses some helper
functions to emulate some ES2015+ features, such as spread syntax. The
TypeScript compiler can be configured to emit these helpers into the
transpiled code (which is controlled by the `noEmitHelpers` option -
false by default). It can also be configured to import these helpers
from the `tslib` module (which is controlled by the `importHelpers`
option - false by default).

While most of the time the helpers will be either emitted or imported,
it is possible that one configures their app to neither emit nor import
them. In that case, the helpers could, for example, be made available on
the global object. This is what `@nativescript/angular`
v9.0.0-next-2019-11-12-155500-01 does. See, for example, [common.js][1].

Ngcc must be able to detect and statically evaluate these helpers.
Previously, it was only able to detect emitted or imported helpers.

This commit adds support for detecting these helpers if they are neither
emitted nor imported. It does this by checking identifiers for which no
declaration (either concrete or inline) can be found against a list of
known TypeScript helper function names.

[1]: https://unpkg.com/browse/@nativescript/angular@9.0.0-next-2019-11-12-155500-01/common.js

gkalpak commented:
Thx for the reviews, @petebacondarwin and @JoostK ✨

mary-poppins commented:
You can preview a46458e at https://pr36418-a46458e.ngbuilds.io/.
